menuconfig.py

```diff
@@ -299,6 +299,159 @@
     "brightpurple": curses.COLOR_MAGENTA + 8,
 }
 
+def rgb_to_6cube(r, g, b):
+    """
+    Take a 3 tuple representing the RGB color and return a 3 tuple
+    representing the index in the xterm 6x6x6 color cube
+
+    Xterm uses a RGB color palette, where the values of each component
+    can be between 0-5 (therefore forming a 6x6x6 cube). The catch is
+    that the mapping between RGB and the 6x6x6 cube is non-linear. The
+    6x6x6 cube index of 0 is mapped to a RGB value of 0. 1-5 are mapped
+    to RGB values of 95 with increments of 40.
+
+    Sources:
+    https://commons.wikimedia.org/wiki/File:Xterm_256color_chart.svg
+    https://github.com/tmux/tmux/blob/master/colour.c
+    """
+    # The formula for converting a single RGB value to the 6cube index
+    # 48 is the middle ground between 0 and 95
+    rgb_to_c6 = lambda x: 0 if x < 48 else max(1, int(round((x - 55) / 40)))
+
+    return rgb_to_c6(r), rgb_to_c6(g), rgb_to_c6(b)
+
+def rgb_from_6cube(r6, g6, b6):
+    """
+    Take a 3 tuple represnting indexes in the xterm 6x6x6 color cube
+    and convert it to a 3 tuple representing the RGB color
+    """
+    # The formula from converting the xterm 6cube index to a RGB value
+    rgb_from_c6 = lambda x: 0 if x == 0 else 55 + x*40
+
+    return rgb_from_c6(r6), rgb_from_c6(g6), rgb_from_c6(b6)
+
+def rgb_to_gray_idx(r, g, b):
+    """
+    Convert an RGB color to an xterm 256-color grayscale color that would
+    represent it well, assuming that it is already close to some shade of gray.
+    This "grayscale candidate" can be compared against the best "color
+    candidate" to find a good xterm color to represent the color.
+    """
+    # Caluclate the luminance (gray intensity) of a color from its R, G, B components
+    # Source: https://stackoverflow.com/questions/596216/formula-to-determine-brightness-of-rgb-color
+    luma = 0.299 * r + 0.587 * g + 0.114 * b
+
+    # Closests index in the grayscale palette, which starts at RGB 0x080808, with
+    # stepping 0x0A0A0A
+    idx = int(round((luma - 0x08) / 0x0A))
+
+    # Clamp the index to 0-23, corresponding to 232-255
+    return max(0, min(idx, 23))
+
+def rgb_from_gray_idx(c):
+    """
+    Convert a grayscale index to its closet single RGB component
+    """
+    g = c * 10 + 8
+    return (g, g, g)
+
+# Obscure Python: rgb2index is initialized from a reference to a global {}
+# Modification to this dictionary are retained between calls rgb_alloc(), thus
+# making rgb2index a static variable
+def rgb_alloc(rgb, rgb2index={}):
+    """
+    Ncurses doesn't allow you to define new colors -- you are allowed only
```

Comment:
Might be nice to start the docstring with a high-level description of what the function does:

    Initialize a new entry in the xterm palette to the given RGB color, returning its
    index. If the color has already been initialized, the index of the existing entry is
    returned.

    Ncurses doesn't allow...